### PR TITLE
fix(grpc): parse target for ipv6 and missing port

### PR DESCRIPTION
**Pull request by majorgreys**
## Description

The target address can lack a port as is the case in [firestore.transports.grpc.FirestoreGrpcTransport](https://github.com/googleapis/python-firestore/blob/e57258c51e4b4aa664cc927454056412756fc7ac/google/cloud/firestore_v1/services/firestore/transports/grpc.py#L210-L253). Being that we look for the right most colon, this would also break any IPv6 addresses. The parsing of the target has been reworked as to properly handle IPv6 addresses.

Blocked on #2384

## Checklist
- [ ] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by brettlangdon — COMMENTED** [review]
test case?

**Review by P403n1x87 — APPROVED** [review]

File: ddtrace/contrib/grpc/client_interceptor.py [hunk 1]
```diff
@@ -180,7 +180,9 @@ def _intercept_client_call(self, method_kind, client_call_details):
         set_grpc_method_meta(span, client_call_details.method, method_kind)
         span._set_str_tag(constants.GRPC_SPAN_KIND_KEY, constants.GRPC_SPAN_KIND_VALUE_CLIENT)
         span._set_str_tag(constants.GRPC_HOST_KEY, self._host)
-        span._set_str_tag(constants.GRPC_PORT_KEY, self._port)
+        if self._port:
+            span._set_str_tag(constants.GRPC_PORT_KEY, str(self._port))
+        span._set_str_tag(constants.GRPC_SPAN_KIND_KEY, constants.GRPC_SPAN_KIND_VALUE_CLIENT)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/grpc/client_interceptor.py:185]:
```suggestion
```

This is set on line 181

File: ddtrace/contrib/grpc/utils.py [hunk 2]
```diff
@@ -19,7 +26,27 @@ def set_grpc_method_meta(span, method, method_kind):
...
     span._set_str_tag(constants.GRPC_METHOD_PATH_KEY, method_path)
-    span._set_str_tag(constants.GRPC_METHOD_PACKAGE_KEY, method_package)
+    if method_package is not None:
+        span._set_str_tag(constants.GRPC_METHOD_PACKAGE_KEY, method_package)
     span._set_str_tag(constants.GRPC_METHOD_SERVICE_KEY, method_service)
     span._set_str_tag(constants.GRPC_METHOD_NAME_KEY, method_name)
     span._set_str_tag(constants.GRPC_METHOD_KIND_KEY, method_kind)
+
+
+def parse_target_from_args(args, kwargs):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/grpc/utils.py:36]:
```suggestion
def _parse_target_from_args(args, kwargs):
```

File: ddtrace/contrib/grpc/utils.py [hunk 3]
```diff
@@ -19,7 +26,27 @@ def set_grpc_method_meta(span, method, method_kind):
...
+    if "target" in kwargs:
+        target = kwargs["target"]
+    else:
+        target = args[0]
+
+    try:
+        if target is not None:
+            parsed = parse.urlsplit("//" + target)
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/contrib/grpc/utils.py]: The URI must follow the spec. I've updated the code to be clear about this. 

```
>>> from ddtrace.internal.compat import parse
>>> parse.urlsplit("localhost:8080")
SplitResult(scheme='', netloc='', path='localhost:8080', query='', fragment='')
>>> parse.urlsplit("//localhost:8080")
SplitResult(scheme='', netloc='localhost:8080', path='', query='', fragment='')
```

**Review by Kyle-Verhoog — APPROVED** [review]
I imagine there are many other places in our library where we don't handle ipv6 properly 😬 

:+1: :+1: for the tests

File: ddtrace/contrib/grpc/utils.py [hunk 4]
```diff
@@ -24,3 +31,26 @@ def set_grpc_method_meta(span, method, method_kind):
     span._set_str_tag(constants.GRPC_METHOD_SERVICE_KEY, method_service)
     span._set_str_tag(constants.GRPC_METHOD_NAME_KEY, method_name)
     span._set_str_tag(constants.GRPC_METHOD_KIND_KEY, method_kind)
+
+
+def _parse_target_from_args(args, kwargs):
```
Comment (jd) [inline ddtrace/contrib/grpc/utils.py:36]: might be cool to have typing

File: ddtrace/contrib/grpc/utils.py [hunk 5]
```diff
@@ -24,3 +31,26 @@ def set_grpc_method_meta(span, method, method_kind):
...
+        # https://tools.ietf.org/html/rfc3986#section-3.2
+        parsed = parse.urlsplit("//" + target if not target.startswith("//") else target)
+        port = None
+        try:
+            port = parsed.port
+        except ValueError:
+            log.warning("Non-integer port in target '%s'", target)
+        return parsed.hostname, port
+    except ValueError:
+        log.warning("Malformed target '%s'.", target)
```
Comment (jd) [inline ddtrace/contrib/grpc/utils.py:56]:
```suggestion
    if target is None:
        return

    try:
        # ensure URI follows RFC 3986 and is preceeded by double slash
        # https://tools.ietf.org/html/rfc3986#section-3.2
        parsed = parse.urlsplit("//" + target if not target.startswith("//") else target)
    except ValueError:
        log.warning("Malformed target '%s'.", target)
        return None

    try:
        port = parsed.port
    except ValueError:
        log.warning("Non-integer port in target '%s'", target)
        port = None
        
    return parsed.hostname, port
```

**Review by jd — APPROVED** [review]
I won't block on my review but it'd be better to have properly scoped try/except
The current one are way too large.

**Comment by Kyle-Verhoog** [comment]
@mergifyio backport 0.48

**Pull request merged**